### ISSUE TITLE
ebpf - cross-compilation support

### ISSUE THREAD

**Issue by waby38b** [netdata/netdata#11594]
##### Bug report summary

Cross compilation of Netdata failed when ebpf is enabled
This failure append during "configure"

"configure.ac" use AC_CHECK_FILE during configure process which is not permitted when cross compiling
(...)
**configure: error: cannot check for file existence when cross compiling**

```
    AC_CHECK_FILE(
        externaldeps/libbpf/libbpf.a,
        [have_libbpf=yes],
        [have_libbpf=no]
    )
```
##### OS / Environment

```
Linux gnbcxl0001 5.11.0-36-generic #40~20.04.1-Ubuntu SMP Sat Sep 18 02:14:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=20.04
```
For cross -compilation, I use buildroot-2021.08.x
=> I want to generate armv7 binaries

##### Netdata version

v1.28 and later

##### Installation method

I use buildroot cross-compilcation process (v2021.08.x)

##### Component Name

"configure", based on configure.ac (autoconf)

##### Steps To Reproduce

1. setup buildroot 2021.08.x
2. enable both libbpf and netdata (--enable-ebpf)
3. make

```
checking for LIBELF... yes
checking for struct bpf_prog_info... yes
checking for externaldeps/libbpf/libbpf.a... configure: error: cannot check for file existence when cross compiling
make[2]: *** [package/pkg-generic.mk:262: /local/home/waby38b/buildroot-2021.08/outputs/build/netdata-1.31.0/.stamp_configured] Error 1
make[1]: *** [Makefile:23: _all] Error 2
make[1]: Leaving directory '/local/home/waby38b/outputs/starkl-custom-cc5119b5/buildroot'
make: *** [GNUmakefile:306: outputs/starkl-custom-cc5119b5/images] Error 2

```
##### Expected behavior

configure process pass without error using standard package detection
I've made a prototype which is working for me both in compilation & cross-compilation
I've replace AC_CHECK_FILE (not supported be cross-compilation) by a standard PKG_CHECK_MODULES

[0001-Use-PKG-detection-for-libbpf-cross-compilation-ok.patch.zip](https://github.com/netdata/netdata/files/7252459/0001-Use-PKG-detection-for-libbpf-cross-compilation-ok.patch.zip)

**Comment by thiagoftsm** [comment]
Hello @waby38b ,

Thank you for your report, I already had a talk with @Ferroin about this reported bug. We are planning to bring eBPF support for ARM during this quarter.

Best regards!

**Comment by waby38b** [comment]
Hi @thiagoftsm ,

Thanks for your msg, but I'm not sure to catch your point.... since I use source of official [libbpf](https://github.com/libbpf/libbpf) library (v0.5.0)... So this is already aligned with your new [release](https://github.com/netdata/libbpf/archive/refs/tags/v0.5_netdata.tar.gz) ?!?

I use [buildroot](http://buildroot.org/) project to build (from scatch) my configuration.

My Linux Kernel is 5.14.y serie (latest stable), also tested with 5.15 on [4kOpen](https://www.4kopen.com/) ARM32 board
Here main lib/version used 
libbpf-0.5.0
llvm-12.0.1
host-clang-12.0.1
clang-12.0.1

I've recompile kernel collector
netdata-kernel-collector-0.8.6

=> if I used the one precompile from [github](https://github.com/netdata/kernel-collector/releases/tag/v0.8.6), I get an error like that:
seem to have some x64 symbol specific
```
$ ./ebpf.plugin 
2021-11-05 17:25:01:  INFO  : MAIN : Does not have a configuration file inside `/etc/netdata/ebpf.d.conf. It will try to load stock file.
2021-11-05 17:25:01:  INFO  : MAIN : Name resolution is disabled, collector will not parser "hostnames" list.
2021-11-05 17:25:01:  INFO  : MAIN : The network value of CIDR 127.0.0.1/8 was updated for 127.0.0.0 .
2021-11-05 17:25:01:  ERROR : MAIN : PROCFILE: Cannot open file '/etc/netdata/apps_groups.conf' (errno 2, No such file or directory)
2021-11-05 17:25:01:  INFO  : MAIN : Cannot read process groups configuration file '/etc/netdata/apps_groups.conf'. Will try '/usr/lib/netdata/conf.d/apps_groups.conf'
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF PROCESS : thread created with task id 735
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : thread created with task id 736
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF PROCESS : set name of thread 735 to EBPF PROCESS
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : set name of thread 736 to EBPF SOCKET
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF CACHESTAT : thread created with task id 737
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF CACHESTAT : set name of thread 737 to EBPF CACHESTAT
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SYNC : thread created with task id 738
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SYNC : set name of thread 738 to EBPF SYNC
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF DCSTAT : thread created with task id 739
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF DCSTAT : set name of thread 739 to EBPF DCSTAT
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF CACHESTAT : thread with task id 737 finished
libbpf: kprobe perf_event_open() failed: No such file or directory
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : Name resolution is disabled, collector will not parser "hostnames" list.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The network value of CIDR 127.0.0.1/8 was updated for 127.0.0.0 .
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (127.0.0.1/8) is inside the range/value (127.0.0.1/8) already inserted, it will be ignored.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (10.0.0.0/8) is inside the range/value (10.0.0.0/8) already inserted, it will be ignored.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (172.16.0.0/12) is inside the range/value (172.16.0.0/12) already inserted, it will be ignored.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (192.168.0.0/16) is inside the range/value (192.168.0.0/16) already inserted, it will be ignored.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (fc00::/7) is inside the range/value (fc00::/7) already inserted, it will be ignored.
2021-11-05 17:25:01: ebpf.plugin INFO  : EBPF SOCKET : The range/value (::1/128) is inside the range/value (::1/128) already inserted, it will be ignored.
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_sync' perf event: No such file or directory
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_syncfs' perf event: No such file or directory
CHART ebpf.file_descriptor '' 'Open and close calls' 'calls/s' 'File' '' 'line' 21000 1
DIMENSION do_sys_open open incremental 1 1
DIMENSION __close_fd close incremental 1 1
CHART ebpf.deleted_objects '' 'Remove files' 'calls/s' 'VFS' '' 'line' 21002 1
DIMENSION vfs_unlink delete incremental 1 1
CHART ebpf.io '' 'Calls to IO' 'calls/s' 'VFS' '' 'line' 21003 1
DIMENSION vfs_read read absolute 1 1
DIMENSION vfs_write write absolute 1 1
CHART ebpf.io_bytes '' 'Bytes written and read' 'bytes/s' 'VFS' '' line 21004 1
DIMENSION vfs_read read absolute 1 1
DIMENSION vfs_write write absolute 1 1
CHART ebpf.process_thread '' 'Start process' 'calls/s' 'Process' '' 'line' 21006 1
DIMENSION _do_fork process absolute 1 1
DIMENSION sys_clone thread absolute 1 1
CHART ebpf.exit '' 'Exit process' 'calls/s' 'Process' '' 'line' 21007 1
DIMENSION do_exit process absolute 1 1
DIMENSION release_task task absolute 1 1
CHART ebpf.process_status '' 'Process not closed' 'difference' 'Process' '' line 21008 1 ''
DIMENSION process '' absolute 1 1
DIMENSION zombie '' absolute 1 1
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_msync' perf event: No such file or directory
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_fsync' perf event: No such file or directory
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_fdatasync' perf event: No such file or directory
libbpf: kprobe perf_event_open() failed: No such file or directory
libbpf: prog 'netdata_syscall_sync': failed to create kprobe '__x64_sys_sync_file_range' perf event: No such file or directory
2021-11-05 17:25:02: ebpf.plugin INFO  : EBPF DCSTAT : thread with task id 739 finished
libbpf: Error in bpf_create_map_xattr(tbl_nv_udp):Invalid argument(-22). Retrying without BTF.
libbpf: load bpf program failed: Permission denied
libbpf: -- BEGIN DUMP LOG ---
libbpf:
``` 
So, if I use my own compiled one, It's ok !

and for netdata, just patch it to used regular libbpf-0.5.0
netdata-1.31.0

I can can provide a Qemu ARM VM already builded if you can have a look....

**Comment by thiagoftsm** [comment]
Hello @waby38b ,

I was thinking that you were using the version `0.4.2` instead the current `0.5.0`. Thanks to bring this update.

Now about the reported problem, as you can see in your error log, there was a try to append a `kprobe` for `__x64_sys_fdatasync`, but this is a function present only on `x86_64` platforms. Since kernel `4.17.0`, kernel team began to append platform as prefix for syscall names. To solve this issue when we are compiling our current eBPF software, I pushed this [commit](https://github.com/netdata/kernel-collector/pull/266/commits/12f73ae2871fc05b7fe28fdfd776fcc69791d3f0) for a PR I am working right now, so instead to use only `X86` platform, we will use the output of `uname -m` command. I expect this PR to be merged in the beginning of next week.

Considering that you reported error for `sync` thread, that uses only syscall, I will give a suggestion for you, a suggestion that probably will show others, because our eBPF programs are compiled right now using a docker image running x64 platforms, but it is also possible that it will work, please, edit your `/etc/netdata/ebpf.d.conf` and change the next threads for no:

```conf
[ebpf programs]
    mount = no
    sync = no
    shm = no
```

These are the threads using `syscalls`.

**Comment by waby38b** [comment]
Hello @thiagoftsm ,

As per your suggestion (disabling mount, sync, shm) , I've now run /usr/libexec/netdata/plugins.d/ebpf.plugin without error
using  precompile BPF kernel-collector from [github](https://github.com/netdata/kernel-collector/releases/download/v0.8.6/netdata-kernel-collector-glibc-v0.8.6.tar.xz) on my ARM32 platform

and here output of "**bpftool prog list**"
```
root@4kOpen-Jeje-host ~
$ /usr/sbin/bpftool prog list
64: tracepoint  name netdata_tracepo  tag 4897571a11319095  gpl
	loaded_at 2021-11-08T14:00:02+0100  uid 0
	xlated 440B  not jited  memlock 4096B  map_ids 29,30,31
	btf_id 43
65: kprobe  name netdata_release  tag cb69b946c019fc45  gpl
	loaded_at 2021-11-08T14:00:02+0100  uid 0
	xlated 448B  not jited  memlock 4096B  map_ids 29,30,31
	btf_id 43
66: tracepoint  name netdata_tracepo  tag 0a2baf337be9abf4  gpl
	loaded_at 2021-11-08T14:00:02+0100  uid 0
	xlated 624B  not jited  memlock 4096B  map_ids 29,30,31
	btf_id 43
67: tracepoint  name netdata_tracepo  tag f3762d47b32c66a6  gpl
	loaded_at 2021-11-08T14:00:02+0100  uid 0
	xlated 992B  not jited  memlock 4096B  map_ids 29,30,31
	btf_id 43
68: kprobe  name netdata_sys_clo  tag 270c2ff80dbdaeb8  gpl
	loaded_at 2021-11-08T14:00:02+0100  uid 0
	xlated 336B  not jited  memlock 4096B  map_ids 30,31
	btf_id 43
```
Here full logs of ebpf.plugin
[ebpf-kernel-collector-v0.8.6-plugins-on-ARM32.txt.zip](https://github.com/netdata/netdata/files/7497078/ebpf-kernel-collector-v0.8.6-plugins-on-ARM32.txt.zip)

But, unfortunately, output of kprobe/tracepoint alway return 0... even with IO activity on my ARM32 setup....
```
BEGIN apps.vfs_write_call
BEGIN apps.vfs_write_bytes
SET vfs_write = 0
SET vfs_write = 0
```

So, it's not (as expected) better with your precompile kernel-collector vs my own compiled from source kernel-collector

I'm open If you have suggestion to debug this ;)

**Comment by thiagoftsm** [comment]
Hello @waby38b ,

I am glad that we give a step forward. :) 
I also observed that some numbers stored for thread and task are pretty high. 
Before I take another action at `kernel-collector` repo, I will require you to run only one more command, please, after to load the eBPF.plugin, can you run the following command:

```bash
# bpftool map dump name tbl_vfs_stats
```

It will print for you all values inside the table:

```
{
        "key": 16,
        "values": [{
                "cpu": 0,
                "value": 1430076
            },{
                "cpu": 1,
                "value": 1448265
            },{
                "cpu": 2,
                "value": 1575320
            },{
                "cpu": 3,
                "value": 1592435
            }
```

Thank you!

**Comment by waby38b** [comment]
Hi @thiagoftsm ,

hum... I think you spot something ;)

With official (or my own compiled) kernel-collector, this return, just nothing....

```
root@4kOpen-Jeje-host ~
$ /usr/sbin/bpftool prog list
109: tracepoint  name netdata_tracepo  tag 4897571a11319095  gpl
	loaded_at 2021-11-09T09:15:14+0100  uid 0
	xlated 440B  not jited  memlock 4096B  map_ids 73,74,75
	btf_id 51
110: kprobe  name netdata_release  tag cb69b946c019fc45  gpl
	loaded_at 2021-11-09T09:15:14+0100  uid 0
	xlated 448B  not jited  memlock 4096B  map_ids 73,74,75
	btf_id 51
111: tracepoint  name netdata_tracepo  tag 0a2baf337be9abf4  gpl
	loaded_at 2021-11-09T09:15:14+0100  uid 0
	xlated 624B  not jited  memlock 4096B  map_ids 73,74,75
	btf_id 51
112: tracepoint  name netdata_tracepo  tag f3762d47b32c66a6  gpl
	loaded_at 2021-11-09T09:15:14+0100  uid 0
	xlated 992B  not jited  memlock 4096B  map_ids 73,74,75
	btf_id 51
113: kprobe  name netdata_sys_clo  tag 270c2ff80dbdaeb8  gpl
	loaded_at 2021-11-09T09:15:14+0100  uid 0
	xlated 336B  not jited  memlock 4096B  map_ids 74,75
	btf_id 51

root@4kOpen-Jeje-host ~
$ /usr/sbin/bpftool map dump name tbl_vfs_stats

root@4kOpen-Jeje-host ~
$ 
```

**Comment by thiagoftsm** [comment]
Hello @waby38b ,

This was exactly what I was thinking.
Today I had a busy day, but tomorrow I will get an ancient 32 bits laptop that I still have, and I will do some tests on it. I installed kernel `5.15.0`, so I am not far from you at least on this. I am expecting at least to repeat one of the problems you are having with it.

**Comment by thiagoftsm** [comment]
Hello @waby38b,

I tested on `i686` computer as I said to you, but the eBPF programs did not load giving error `4007`. According some information I read in other forums, this has relationship with some features we are using. I will dig more this error during next days.

Best regards!

**Comment by waby38b** [comment]
Hi @thiagoftsm ,

Good News !!! 
Just tested [netdata v1.32.0](https://github.com/netdata/netdata/releases/tag/v1.32.0) and the just just released [libbpf-.0.6.0](https://github.com/libbpf/libbpf/releases/tag/v0.6.0) on my ARM32 platform
...and now, it work PERFECTLY !! (using my own compiled BPF kernel-collector)

I do not know (yet!) if this is due to netdata-v1.32.0 or/and libbpf-0.6.0, but it work !!

And as expected, new this tool return some values: (was previously broken)
```
root@4kOpen-Jeje-host ~
$ /usr/sbin/bpftool map dump name tbl_vfs_stats
[{
        "key": 0,
        "values": [{
                "cpu": 0,
                "value": 6922
            },{
                "cpu": 1,
                "value": 166830
            },{
                "cpu": 2,
                "value": 6461
            },{
                "cpu": 3,
                "value": 11117
            }
        ]
    },{
        "key": 1,
        "values": [{
                "cpu": 0,
                "value": 0
            },{
(...)
```

and I have some nice graphs view under my webbrowser :)

So, to summarize, 
My Linux Kernel is 5.15.y serie (latest stable), [4kOpen](https://www.4kopen.com/) ARM32 board
Here main lib/version used
```
libbpf-0.6.0
llvm-12.0.1
host-clang-12.0.1
clang-12.0.1
netdata-v1.32.0
```
for netdata, I still use my [patch](https://github.com/netdata/netdata/files/7252459/0001-Use-PKG-detection-for-libbpf-cross-compilation-ok.patch.zip) for ebpf - cross compilation support 
and, also, still need recompile kernel collector (since precompiled still have x64 symbol)
`netdata-kernel-collector-0.8.6
`

I've seen from git that kernel collector will use CO-RE, I wait for net release ;)

![image](https://user-images.githubusercontent.com/44431134/144414640-d95eabc9-cbe6-41e4-a6d1-c0bdead1cafb.png)

**Comment by thiagoftsm** [comment]
Awesome news @waby38b ! :rofl: 
I am planning to bring `libbpf-0.6` to netdata as soon our code freeze ends. 

I am closing the issue, because you had success, but be free to reopen it if is necessary.

Thank you very much for your feedback!
